feat: Creates design doc for enabling access to newly created hts token add… (#3258)
hashgraph/hedera-json-rpc-relay: docs/design/hts_address_tx_receipt.md

# HTS Token Creation via SystemContract Receipt Enhancement

## Purpose
The purpose of this enhancement is to improve the interoperability between Hedera Token Service (HTS) and Ethereum-compatible tools, like ethers, by ensuring access to the necessary information about newly created tokens, in this case token address. This change specifically addresses the gap in Web3 tooling support when creating tokens through HTS precompile contracts, making user experience smoother.

## Problem Statement

> When users create tokens (fungible and non-fungible) by directly calling the 0x167 contract there is no way for to get the contract address of the newly created token via ethers. This is because the transaction receipt's contractAddress field is not being populated with the HTS token address, since essentially the operations is just a contract call, but not a contract create (deploy).

Reviewer: ```suggestion
When users create tokens (fungible and non-fungible) by directly calling the 0x167 contract there is no way for them to get the evm address of the newly created token via standard EVM tooling e.g. ethers. This is because the transaction receipt's contractAddress field is not being populated with the HTS token address, since essentially the operations is just a contract call, but not a contract create (deploy).
```

Author: Done.

## Current Behaviour

Currently, when creating a token via HTS the address of the system contract is returned as a contractAddress in the transaction receipt.

## Proposed Solution

### Technical Details

1. Function Signature Detection

> Detect in the transactionResponse from the mirror node, if the transaction was calling any of the HTS method e.g createFungibleToken/createNonFungibleToken etc.

Reviewer: Please list all current function that this intends to support

N.B Currently, HTS supports both v1 and v2 security model function selectors

> 1. Extract the token address from the call_result field in the transaction response.

Reviewer: Please lease an example of a call_result and what is looks like before and after extraction

> 1. Extract the token address from the call_result field in the transaction response.

Reviewer: What approach are you intending to utilize to extract?
Is it just brute force string manipulation or are you using additional logic to parse the bytes into an expected format and pull the address from that?

Author: Brute force manipulation I would say, Improved the description

2. Add the token address to the receipt

> 

Reviewer: Are there existing interfaces or higher level classes of not that should be modified to support this.
Essentially you should be telling us what high level changes to the code base you are suggesting

Author: I dont't think we need to change any higher level classes. Only the getTransactionReceipt method in eth.ts file
So not sure what else should I add in terms of this note?

Reviewer: Sounds good

## Testing Requirements

> 1. Unit tests

Reviewer: UT's aren't necessary in the design doc.
Acceptance tests highlighting the difference scenarios is what should be highlighted.

   1. Test token creation function signature detection

2. Acceptance tests

>    1. Test fungible token creation flow via ethers and correct receipt response

Reviewer: There should be a test case for every create method being supported.

Author: Yes, of course
I just didnt list them in detail, thought its not necessary, but will do it now :) its good to have a detailed test plan

## Problem Statement
When users create tokens (fungible and non-fungible) by directly calling the 0x167 contract there is no way for them to get the evm address of the newly created token via standard EVM tooling e.g ethers. This is because the transaction receipt's contractAddress field is not being populated with the HTS token address, since essentially the operations is just a contract call, but not a contract create (deploy).

## Current Behaviour

Currently, when creating a token via HTS the address of the system contract is returned as a contractAddress in the transaction receipt.

## Proposed Solution

### Technical Details

### 1. Function Signature Detection

Detect in the transactionResponse from the mirror node, if the transaction was calling any of the HTS methods - createFungibleToken/createNonFungibleToken/createFungibleTokenWithCustomFees/createNonFungibleTokenWithCustomFees

> N.B Currently, HTS supports both v1 and v2 security model function selectors

Reviewer: Suggestion to be more specific. Also it's not security model related
```suggestion
N.B The following forms of the [HTS system contract](https://github.com/hashgraph/hedera-smart-contracts/tree/v0.10.1/contracts/system-contracts/hedera-token-service) create token function selectors will be supported. This is as of the 0.56 version of the services.
```

Reviewer 2: qq: I see Nana's suggestion above was marked as resolved but didn't see the changes nor updates. What's the approach here?

Author: yes, the commit for some reason got lost, because i didnt pull it locally and then forced push

Reviewer 2: Okay thanks for confirming!

```javascript
const HTS_CREATE_FUNCTIONS_SIGNATURE = [
  "createFungibleToken(...)",
  "createNonFungibleToken(...)",
  "createFungibleTokenWithCustomFees(...)",
  "createNonFungibleTokenWithCustomFees(...)"
];
const functionSelector = receiptResponse.function_parameters.substring(0, FUNCTION_SELECTOR_CHAR_LENGTH);
const isTokenCreation = HTS_CREATE_FUNCTIONS_SIGNATURE.some(signature => 
  Utils.calculateFunctionSelector(signature) === functionSelector
);
```

> ### 2. Extract the token address from the call_result field in the transaction response. The extractions can happen with brute force since we know what the function response is beforehand and we know the last 40 characters of the response are the token address.

Reviewer: ```suggestion
### 2. Extract the token address from the `call_result` field in the transaction response. 
The extractions can performed with string matching logic since we know what the function response is beforehand and we know the last 40 characters of the response are the token address.
```

Reviewer 2: same here suggestion was marked resolve but no updates nor comments

# HTS Token Creation via SystemContract Receipt Enhancement

## Purpose
The purpose of this enhancement is to improve the interoperability between Hedera Token Service (HTS) and Ethereum-compatible tools, like ethers, by ensuring access to the necessary information about newly created tokens, in this case token address. This change specifically addresses the gap in Web3 tooling support when creating tokens through HTS precompile contracts, making user experience smoother.

## Problem Statement
When users create tokens (fungible and non-fungible) by directly calling the 0x167 contract there is no way for them to get the evm address of the newly created token via standard EVM tooling e.g ethers. This is because the transaction receipt's contractAddress field is not being populated with the HTS token address, since essentially the operations is just a contract call, but not a contract create (deploy).

## Current Behaviour

Currently, when creating a token via HTS the address of the system contract is returned as a contractAddress in the transaction receipt.

## Proposed Solution

> ### Technical Details

Reviewer 2: Is it possible to mention somewhere the endpoints are being modified for this logic?

Author: Well, this affects the getTransactionReceipt endpoint only
As it is mentioned in a few places that we want to return it there?
Could you suggest a specific place?

Reviewer 2: Ahh okay I see. I would suggest creating a dedicated section to clarify that the feature focuses on specific endpoint. As when I read through the doc, I was wondering what endpoints are being focused. So perhaps we could add a section and title it "Affected Endpoints," "Impacted Endpoints," "Targeted Endpoints," or something like that.

> ### 1. Function Signature Detection

Reviewer: Notably, this should be done in a way that can be generalized to other future system contracts e.g. createSchedule or createTopic

Author: Well, i have shown in the code block below that my idea is to have an array with the function signatures and with the help of it we get the function selector. So in the future any new system contract methods can be easily added. Is this what you mean?

> Detect in the transactionResponse from the mirror node, if the transaction was calling any of the HTS methods - createFungibleToken/createNonFungibleToken/createFungibleTokenWithCustomFees/createNonFungibleTokenWithCustomFees

Reviewer 2: `if the transaction was calling any of the HTS methods` this could be a little misleading as the listed methods are only about token creation. 
```suggestion
Detect in the transactionResponse from the mirror node, if the transaction was calling any of the token creation HTS methods - createFungibleToken/createNonFungibleToken/createFungibleTokenWithCustomFees/createNonFungibleTokenWithCustomFees
```

Author: I think the purpose of this sentence is to name the specific methods in the end.
If you really want it to be clear i can change it to
-
```
Detect in the transactionResponse from the mirror node, if the transaction was calling any of the FOLLOWING HTS methods - createFungibleToken/createNonFungibleToken/createFungibleTokenWithCustomFees/createNonFungibleTokenWithCustomFees
```
As Nana mentioned in another comment this may change to include other methods as well, so i am reluctant to specifying if they are token creation or anything else. What do you think

Reviewer 2: Nice! I like it! Thanks for the update!

N.B Currently, HTS supports both v1 and v2 security model function selectors

```javascript
const HTS_CREATE_FUNCTIONS_SIGNATURE = [
  "createFungibleToken(...)",
  "createNonFungibleToken(...)",
  "createFungibleTokenWithCustomFees(...)",
  "createNonFungibleTokenWithCustomFees(...)"

> ];

Reviewer 2: If possible, let's include the correct params for these methods instead of `(...)`

const functionSelector = receiptResponse.function_parameters.substring(0, FUNCTION_SELECTOR_CHAR_LENGTH);
const isTokenCreation = HTS_CREATE_FUNCTIONS_SIGNATURE.some(signature => 
  Utils.calculateFunctionSelector(signature) === functionSelector
);
```

### 2. Extract the token address from the `call_result` field in the transaction response. 
The extractions can performed with string matching logic since we know what the function response is beforehand and we know the last 40 characters of the response are the token address.

Example call_result. First 64 characters excluding 0x prefix are the response code, second 64 characters are the token address:
```json
"call_result": "0x0000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000000000000000040a"
```

```javascript
const tokenAddress = receiptResponse.call_result.substring(receiptResponse.call_result.length - 40);
```
After extraction, the expected output is:
```
"0x000000000000000000000000000000000000000000000000000000000000040a"
```

> ### 3. Add the token address to the receipt

Reviewer: nit: nothing in this section.
Did you mean to highlight that you are replacing the default `0x167` value with this?

# HTS Token Creation via SystemContract Receipt Enhancement

## Purpose
The purpose of this enhancement is to improve the interoperability between Hedera Token Service (HTS) and Ethereum-compatible tools, like ethers, by ensuring access to the necessary information about newly created tokens, in this case token address. This change specifically addresses the gap in Web3 tooling support when creating tokens through HTS precompile contracts, making user experience smoother.

## Problem Statement
When users create tokens (fungible and non-fungible) by directly calling the 0x167 contract there is no way for them to get the evm address of the newly created token via standard EVM tooling e.g ethers. This is because the transaction receipt's contractAddress field is not being populated with the HTS token address, since essentially the operations is just a contract call, but not a contract create (deploy).

## Current Behaviour

Currently, when creating a token via HTS the address of the system contract is returned as a contractAddress in the transaction receipt.

## Proposed Solution

> ### Targeted endpoints
> 
> [getTransactionReceipt]

Reviewer 2: Perhaps below.
```suggestion
### Targeted endpoints

- eth_getTransactionReceipt
```
